### PR TITLE
Gamemode System: Group Objectives

### PR DESCRIPTION
**Pull request by Ryan089**
## Summary

This PR allows for group objectives to be created by setting the appropriate Collaboration Type on the objective definition:
- **Individual**: relevant only to the player.
- **Cooperative**: once one player succeeds, all other assignees succeed.
- **Competitive**: once one player succeeds, all other assignees fail.

## Changes to Files

- **GamemodeTests.cs** added new tests to confirm the competitive and cooperative objective functionality, and confirm late join functionality. Minor tweaks to other tests were included to aid maintainability.
- **GamemodeObjective.cs** added fields for the collaboration type, and amount of objective assignees. *Note: Alignment requirement is currently unused*.
- **Gamemode.cs** separated the creation of round start objectives from the assignment of those objectives, and manages finalization of shared objectives.
- **ObjectiveFactory.cs** allows easier access to objective creation and duplication.
- **GetItemObjective.cs** replaces **GetNukeCardObjective.cs** by providing a generic means of creating objectives where a specific item must be found.
- **GamemodeObjectiveItemView.cs** changes the view colours to distinguish "In Progress" objectives (yellow) from "Failed" objectives (red).

### CODE REVIEW

File: Assets/Scripts/EditorTests/GamemodeTests.cs [hunk 1]
```diff
@@ -214,18 +212,75 @@ public void FinalizeGamemodeDoesNotChangeStatusOfCompletedObjectives(Gamemode su
...
         }
 
+        /// <summary>
+        /// Test to confirm that when one player succeeds in a multi-participant cooperative objective, all other players
+        /// involved in that objective also succeed.
+        /// </summary>
+        [Test]
+        public void CooperativeObjectivesSucceedWhenOnePlayerCompletesThem()
+        {
+
```
Comment (joaoburatto) [inline Assets/Scripts/EditorTests/GamemodeTests.cs]: remove blank space

File: Assets/Scripts/EditorTests/GamemodeTests.cs [hunk 2]
```diff
@@ -214,18 +212,75 @@ public void FinalizeGamemodeDoesNotChangeStatusOfCompletedObjectives(Gamemode su
...
+        }
+
+        /// <summary>
+        /// Test to confirm that when one player succeeds in a multi-participant cooperative objective, all other players
+        /// involved in that objective also succeed.
+        /// </summary>
+        [Test]
+        public void CompetitiveObjectivesFailWhenOnePlayerCompletesThem()
+        {
+
```
Comment (joaoburatto) [inline Assets/Scripts/EditorTests/GamemodeTests.cs]: remove blank space

File: Assets/Scripts/EditorTests/GamemodeTests.cs [hunk 3]
```diff
@@ -214,18 +212,75 @@ public void FinalizeGamemodeDoesNotChangeStatusOfCompletedObjectives(Gamemode su
...
         /// <summary>
         /// Test to confirm that all objectives are cleared when the Gamemode is reset.
         /// </summary>
-        /// <param name="sut">Gamemode (acquired from TestCaseSource) as the System Under Test</param>
         [Test]
-        [TestCaseSource(nameof(AllGamemodes))]
-        public void ResetGamemodeClearsAllObjectives(Gamemode sut)
+        public void ResetGamemodeClearsAllObjectives()
         {
```
Comment (joaoburatto) [inline Assets/Scripts/EditorTests/GamemodeTests.cs]: remove blank space

File: Assets/Scripts/EditorTests/GamemodeTests.cs [hunk 4]
```diff
@@ -236,6 +291,28 @@ public void ResetGamemodeClearsAllObjectives(Gamemode sut)
             Assert.IsTrue(sut.RoundObjectives.Count == 0);
         }
 
+        /// <summary>
+        /// Test to confirm that late joining players also get objectives.
+        /// </summary>
+        [Test]
+        public void LateJoiningPlayersGetObjectives()
+        {
+
```
Comment (joaoburatto) [inline Assets/Scripts/EditorTests/GamemodeTests.cs]: remove blank space

File: Assets/Scripts/EditorTests/GamemodeTests.cs [hunk 5]
```diff
@@ -326,6 +403,27 @@ private List<string> SampleCkeys(int numberOfPlayers)
...
+            collectionEntry.RemainingAssignments = 1000;
+
+            GamemodeObjectiveCollection collection = new GamemodeObjectiveCollection();
+            collection.Entries = new List<GamemodeObjectiveCollectionEntry>();
+            collection.Entries.Add(collectionEntry);
+
+            gamemode.PossibleObjectives = collection;
+
+            return gamemode;
+
```
Comment (joaoburatto) [inline Assets/Scripts/EditorTests/GamemodeTests.cs]: remove blank space

File: Assets/Scripts/SS3D/Systems/Gamemodes/GamemodeObjective.cs [hunk 6]
```diff
@@ -15,12 +15,62 @@ public class GamemodeObjective : ScriptableObject, IGamemodeObjective
...
+        /// Constructor allowing creation of specific objective data.
+        /// </summary>
+        /// <param name="title">Title of the objective (visible to player)</param>
+        /// <param name="collaborationType">Whether the objective is individual, competitive or cooperative</param>
+        /// <param name="alignment">Whether the objective is valid for antagonists, non-antagonists or both</param>
+        /// <param name="minAssignees">Minimum number of players required for this objective</param>
+        /// <param name="maxAssignees">Maximum number of players required for this objective</param>
+        public GamemodeObjective(string title, CollaborationType collaborationType, Alignment alignment, int minAssignees, int maxAssignees)
+        {
+            this._title = title;
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Gamemodes/GamemodeObjective.cs]: remove all this.variable, if the variable is private, it'll have the _ on it, so there's no reason to use this.variable.

File: Assets/Scripts/SS3D/Systems/Gamemodes/Modes/Gamemode.cs [hunk 7]
```diff
@@ -140,35 +146,98 @@ protected virtual void CreateObjectives(List<string> spawnedPlayersCkeys)
...
+            AssignRoundStartObjectives(spawnedPlayersCkeys);
+        }
+
+        /// <summary>
+        /// Very basic implementation of objective assignment. This method simply assigns 
+        /// pre-generated objectives sequentially to the list of Ckeys. Can be overridden for
+        /// custom objective assignment.
+        /// </summary>
+        /// <param name="CKeysToAssign">List of Ckeys to assign objectives to</param>
+        protected virtual void AssignRoundStartObjectives(List<string> CKeysToAssign)
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Gamemodes/Modes/Gamemode.cs]: CkeysToAssign should start with a lower letter, as in "ckeyToAssign"

File: Assets/Scripts/SS3D/Systems/Gamemodes/Modes/Gamemode.cs [hunk 8]
```diff
@@ -140,35 +146,98 @@ protected virtual void CreateObjectives(List<string> spawnedPlayersCkeys)
...
+            }
+
+            // Sequentially assign objectives
+            for (int i = 0; i < CKeysToAssign.Count; i++)
+            {
+                AssignObjective(RoundObjectives[i], CKeysToAssign[i]);
+            }
+        }
+
+        private void CreateRoundStartObjectives(int numberToCreate)
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Gamemodes/Modes/Gamemode.cs]: rename numberToCreate to amountToCreate

File: Assets/Scripts/SS3D/Systems/Gamemodes/Modes/Gamemode.cs [hunk 9]
```diff
@@ -223,7 +289,45 @@ private void FailOnGoingObjectives()
...
+        /// <summary>
+        /// Finalizes any objective that is shared with other players (whether co-opererative or competitive)
+        /// once one of the players has triggered completion.
+        /// </summary>
+        /// <param name="triggerObjective">The objective that was completed</param>
+        private void FinalizeSharedObjective(GamemodeObjective triggerObjective)
+        {
+            // Finalization is only valid if the trigger objective has succeeded, and it is a shared objective. Exit early if invalid.
+            if (triggerObjective.CollaborationType == CollaborationType.Individual) return;
+            if (!triggerObjective.Succeeded) return;
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Gamemodes/Modes/Gamemode.cs]: join both ifs

File: Assets/Scripts/SS3D/Systems/Gamemodes/ObjectiveFactory.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
+using UnityEngine;
+
+namespace SS3D.Systems.Gamemodes
+{
+    /// <summary>
+    /// This class is used to create and duplicate GamemodeObjectives so that the underlying ScriptableObjects are not modified.
+    /// </summary>
+    public static class ObjectiveFactory
+    {
+
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Gamemodes/ObjectiveFactory.cs]: remove blank space

**Review by joaoburatto — CHANGES_REQUESTED** [review]

**Review by TaylorNAlbarnaz — CHANGES_REQUESTED** [review]
Clients can't complete objectives
Clients also don't have their objectives set as complete when the hosts completes them, though they might be user-specific and not shared here, but they should fail if that's the case

**Review by TaylorNAlbarnaz — APPROVED** [review]
lgtm

**Pull request merged**